api/main.py

```diff
@@ -119,6 +119,18 @@ async def put_node(node_id: str, node: Node, token: str = Depends(get_user)):
     return obj
 
 
+@app.get('/get_root_node/{node_id}', response_model=Node)
+async def get_root_node(node_id: str):
+    while True:
+        node = await db.find_by_id(Node, node_id)
+        if node is None:
+            break
+        node_id = node.parent
+        if node_id is None:
+            break
```

Comment:
I think we could change it something like that. It's just a bit more concise. Having said that, this code requires Python 3.8+


```suggestion
    while (node := await db.find_by_id(Node, node_id)) is not None:
        node_id = node.parent
        if node_id is None:
            break
```

Reply:
Shouldn't the loop be doing a `while node_id is not None`? That's the actual condition for ending the loop, then there's no parent so no node_id. If node is None, that's more like an error if an id is not valid.

Reply:
Yes, `if node_id is not None` is the condition until we need to loop through the nodes. In case of invalid `node_id`, we need to catch an `InvalidId` exception. When `node_id` is valid but no node matching with the id exists, we will get `null` output.

Reply:
OK, by "invalid id" I meant an id that doesn't match any node. That should not happen, unless there's an error in the database or a corrupt id. So it should be handled as an error case, but the actual logic is to keep looping as long as there is a node id. Something like:
```
while node_id:
  node = db.get_node(node_id)
  node_id = node.parent
```
and everything else is error handling (i.e. node being None). We could add a check in the middle to `break` if node is None, or catch exceptions around the loop to better separate the actual logic of the code from error handling.